src/Infrastructure.Styles.Analyzer/RMSYN1103ExtensionMethodsTestedAsNormalInvocationCodeFixProvider.cs

```diff
@@ -0,0 +1,145 @@
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Infrastructure.Styles.Analyzer
+{
+    public class RMSYN1103ExtensionMethodsTestedAsNormalInvocationCodeFixProvider : CodeFixProvider
+    {
+        public override FixAllProvider? GetFixAllProvider ()
+        {
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public override Task RegisterCodeFixesAsync (CodeFixContext context)
+        {
+            context.RegisterCodeFix(
+                CodeAction.Create(
+                    "Fix Extension method invocation",
+                    c => CodeFix(c, context),
+                    RMSYN1103ExtensionMethodsTestedAsNormalInvocationSyntaxAnalyzer.DiagnosticId),
+                context.Diagnostics.First());
+            return Task.CompletedTask;
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } =
+            ImmutableArray.Create(RMSYN1103ExtensionMethodsTestedAsNormalInvocationSyntaxAnalyzer.DiagnosticId);
+
+        private static async Task<Document> CodeFix (CancellationToken cancellationToken, CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
+            var diagnosticSpan = diagnostic.Location.SourceSpan;
+
+            var syntaxRoot = await context.Document.GetSyntaxRootAsync(cancellationToken);
+            if (syntaxRoot == null) return context.Document;
+
+            var oldNode = syntaxRoot.FindNode(diagnosticSpan).FirstAncestorOrSelf<InvocationExpressionSyntax>();
+            if (oldNode == null) return context.Document;
+
+            var semanticModel = await context.Document.GetSemanticModelAsync(cancellationToken);
+            if (semanticModel == null) return context.Document;
+
+            var symbolInfo = semanticModel.GetSymbolInfo(oldNode, cancellationToken);
+            if (symbolInfo.Symbol is not IMethodSymbol methodSymbol)
+                return context.Document;
+
+            var containingType = methodSymbol.ContainingType;
+            if (containingType == null) return context.Document;
+
+            var newSyntaxRoot = syntaxRoot.ReplaceNode(
+                oldNode, 
+                TransformExtensionInvocationNode(
+                    oldNode, 
+                    containingType, 
+                    semanticModel)
+                );
+            
+            return context.Document.WithSyntaxRoot(newSyntaxRoot);
```

Comment:
I think we can reduce the amount of if checks here. We can assume that the diagnostic is located at the correct place (at an InvocationExpression). We also should not need the semantic model to replace the node in the tree. That should be possible using the syntactical model alone.

Reply:
Yeah fixing the ifs is fine, but I havent found any way of doing this without semantic model. I cant seem to be able to locate where the definition for the extension method is.

Reply:
We do not need the sematic model for two things:

 1. Checking if a method is an extension method
 2. Getting the type an extension method is referring to

Any other transformations should be done using the syntax tree only.

Reply:
Getting the type an extension method is referring to is exactly the thing I need here, though? Or am I missing something

Reply:
Yes you are correct my comment about the semantic version is a leftover of the time before I recognized that we need the semantic model to determine if a method is an extension method.